platform/ui/src/contextProviders/ViewportGridProvider.tsx

```diff
@@ -90,53 +101,61 @@ export function ViewportGridProvider({ children, service }) {
 
+      /**
+       * Sets the display sets for multiple viewports.
+       * This is a replacement for the older set display set for viewport (single)
+       * because the old one had a race conditions wherein the viewports could
+       * render partially in various ways causing exceptions.
+       */
+      case 'SET_DISPLAYSETS_FOR_VIEWPORTS': {
+        const { payload } = action;
         const viewports = state.viewports.slice();
 
-        let newView = {
-          ...viewport,
-          displaySetInstanceUIDs,
-          viewportOptions,
-          displaySetOptions,
-          viewportLabel: viewportLabels[viewportIndex],
-        };
-        viewportOptions.presentationIds = getPresentationIds(
-          newView,
-          viewports
-        );
-
-        // Make sure we assign a viewport id
-        newView = reuseViewport({}, newView, state.viewports);
-        console.log(
-          'Creating new viewport',
-          viewportIndex,
-          newView.viewportOptions.viewportId,
-          displaySetInstanceUIDs,
-          displaySetOptions
-        );
-
-        viewports[viewportIndex] = newView;
+        // Don't reuse any viewports here.
+        const idSet = state.viewports.reduce((accumulator, viewport) => {
+          accumulator[viewport.viewportOptions.viewportId] = true;
+          return accumulator;
+        }, {});
+
+        for (const updatedViewport of payload) {
+          // Note: there should be no inheritance happening at this level,
+          // we can't assume the new displaySet can inherit the previous
+          // displaySet's or viewportOptions at all. For instance, dragging
+          // and dropping a SEG/RT displaySet without any viewportOptions
+          // or displaySetOptions should not inherit the previous displaySet's
+          // which might have been a PDF Viewport. The viewport itself
+          // will deal with inheritance if required. Here is just a simple
+          // provider.
+          const { viewportIndex, displaySetInstanceUIDs } = updatedViewport;
+          const previousViewport = viewports[viewportIndex] || {};
+          const viewportOptions = { ...updatedViewport.viewportOptions };
+
+          const displaySetOptions = updatedViewport.displaySetOptions || [];
+          if (displaySetOptions.length === 0) {
+            // Only copy index 0, as that is all that is currently supported by this
+            // method call.
+            displaySetOptions.push({
+              ...previousViewport.displaySetOptions?.[0],
+            });
+          }
+
+          let newView = {
+            ...previousViewport,
+            displaySetInstanceUIDs,
+            viewportOptions,
+            displaySetOptions,
+            viewportLabel: viewportLabels[viewportIndex],
+          };
+          viewportOptions.presentationIds = getPresentationIds(
+            newView,
+            viewports
+          );
+
+          newView = reuseViewport(idSet, newView, state.viewports);
```

Comment:
I'm still not clear on what `reuseViewport` do, I guess it finds a reusable viewport right? so first of all `getReusableViewport`? since `reuseViewport` is action
Otherwise, the `reuseViewport` function can benefit a lot from more function documentation + in-line doc